src/App.tsx

```diff
@@ -15,47 +16,120 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  NONE,
+  ALPABET,
+  LENGTH,
+}
+
+type ReorderOptions = {
+  sortType: SortType,
+  isReversed: boolean,
+};
+
+export function getReorderedGoods(
+  goods: string[],
+  { sortType, isReversed }: ReorderOptions,
+) {
+  const visibleGoods = [...goods];
+
+  visibleGoods.sort((previous, current) => {
+    switch (sortType) {
+      case SortType.ALPABET:
+        return previous.localeCompare(current);
+
+      case SortType.LENGTH:
+        return previous.length - current.length;
+
+      default:
+        return 0;
+    }
+  });
+
+  if (isReversed) {
+    visibleGoods.reverse();
+  }
+
+  return visibleGoods;
+}
+
 export const App: React.FC = () => {
+  const [isReversed, setIsReversed] = useState(false);
+  const [sortType, setSortType] = useState(SortType.NONE);
+
+  const goods = getReorderedGoods(
+    goodsFromServer,
+    { sortType, isReversed },
+  );
+
+  const sortByAlphabet = () => setSortType(SortType.ALPABET);
+  const sortByLength = () => setSortType(SortType.LENGTH);
+  const toggleReverse = () => setIsReversed(!isReversed);
```

Comment:
If you are based on the previous value, then you need to use functional set state

Reply:
> If you are based on the previous value, then you need to use functional set state

Initial values of state were located in useState initial values. We use "set state" through these methods (as setSortType, setIsReversed) and they look for initial value in its part of state to change it.
useState(initialState) keeps these values for us. Or may be, I didn't understand the comment.

Reply:
> If you are based on the previous value, then you need to use functional set state

Done!